dll/cpl/sysdm/environment.c

```diff
@@ -144,16 +144,13 @@ GatherDataFromListView(HWND hwndListView,
         dwValueLength += _tcslen(szData);
     }
 
-    /* Reallocate the value buffer, regrowing it if necessary */
-    if (!VarData->lpRawValue || (_tcslen(VarData->lpRawValue) < dwValueLength))
-    {
-        if (VarData->lpRawValue)
-            GlobalFree(VarData->lpRawValue);
+    /* Reallocate the value buffer */
+    if (VarData->lpRawValue)
+        GlobalFree(VarData->lpRawValue);
 
-        VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
-        if (!VarData->lpRawValue)
-            return 0;
-    }
+    VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
+    if (!VarData->lpRawValue)
+        return 0;
```

Comment:
So, what was the fundamental reason why the previous code didn't work?

Reply:
@HBelusca The reason was that since, if the length of changed environment values are of same or smaller than previous one then this will not reallocate the buffer hence if any new value will be tried to get copied to it, it will outright fail.
This was especially a problem when you either deleted the value(s) or even just changed the order of it.

Reply:
> this will not reallocate the buffer

Indeed that was the original intent.

> hence if any new value will be tried to get copied to it, it will outright fail.

That's the actual problem: why would it fail?

Reply:
> > this will not reallocate the buffer
> 
> Indeed that was the original intent.
> 
> > hence if any new value will be tried to get copied to it, it will outright fail.
> 
> That's the actual problem: why would it fail? I would expect that one would just copy the new data (of smaller size, in the case the buffer didn't need to be reallocated) by replacing whatever is already present in the existing buffer (since we don't care about it anyway: now with your PR you just directly free it and replace by a fresh one).

Since the way the values are getting copied is by taking the values from the listview control one to a temporary buffer and then copying it to the actual buffer one by one since it is not possible to directly take all the values from it.